### PR TITLE
Refactor segment allocate policy on querycoord

### PR DESCRIPTION
**Pull request by xige-16**
fix issue: #9481
Signed-off-by: xige-16 <xi.ge@zilliz.com>

### CODE REVIEW

File: internal/querycoord/segment_allocator.go [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+// shuffleSegmentsToQueryNode shuffle segments to online nodes
+// returned are noded id for each segment, which satisfies:
+//     len(returnedNodeIds) == len(segmentIDs) && segmentIDs[i] is assigned to returnedNodeIds[i]
+func shuffleSegmentsToQueryNode(ctx context.Context, collectionID UniqueID, reqs []*querypb.LoadSegmentsRequest, cluster Cluster, wait bool, excludeNodeIDs []int64) error {
+	if len(reqs) == 0 {
+		return nil
+	}
+
+	maxNumSegments := 0
```
Comment (xiaofan-luan) [inline internal/querycoord/segment_allocator.go:45]: We can have two heaps.
1st heap is min heap, 2nd heap is maxheap, always compare two heap top and see if we can move one segment to make the cluster balancer.

File: internal/querycoord/segment_allocator.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+				}
+			}
+		}
+		if lastOffset == offset {
+			loopAll = true
+		}
+	}
+}
+
+func shuffleSegmentsToQueryNodeV2(ctx context.Context, collectionID UniqueID, reqs []*querypb.LoadSegmentsRequest, cluster Cluster, wait bool, excludeNodeIDs []int64) error {
```
Comment (xiaofan-luan) [inline internal/querycoord/segment_allocator.go:106]: similar heap logic here, the heap size can be changed to memory consumption

File: internal/querycoord/channel_allocator.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		if numChannels > maxNumChannels {
+			maxNumChannels = numChannels
+		}
+	}
+	if len(reqs) == 0 {
+		return nil
+	}
+
+	offset := 0
+	loopAll := false
```
[the remark this answers is not in the document]
Comment (xige-16) [reply internal/querycoord/channel_allocator.go:66]: > Why do we find node with least segment in each round and do assignment?

According to the number of dmchanenl already watched on the node to allocate a new channel, it has nothing to do with the segment

File: internal/querycoord/segment_allocator.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/querypb"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+)
+
+const MaxMemUsagePerNode = 0.9
```
Comment (xiaofan-luan) [inline internal/querycoord/segment_allocator.go:27]: has to be a param rather than a constant

**Pull request merged**